src/phoenix/server/api/dataloaders/record_counts.py

```diff
@@ -0,0 +1,102 @@
+from collections import defaultdict
+from datetime import datetime
+from typing import (
+    Any,
+    AsyncContextManager,
+    Callable,
+    DefaultDict,
+    List,
+    Literal,
+    Optional,
+    Tuple,
+)
+
+from sqlalchemy import Select, func, select
+from sqlalchemy.ext.asyncio import AsyncSession
+from strawberry.dataloader import DataLoader
+from typing_extensions import TypeAlias, assert_never
+
+from phoenix.db import models
+from phoenix.server.api.input_types.TimeRange import TimeRange
+from phoenix.trace.dsl import SpanFilter
+
+Kind: TypeAlias = Literal["span", "trace"]
+ProjectRowId: TypeAlias = int
+TimeInterval: TypeAlias = Tuple[Optional[datetime], Optional[datetime]]
+FilterCondition: TypeAlias = Optional[str]
+SpanCount: TypeAlias = int
+
+Segment: TypeAlias = Tuple[Kind, TimeInterval, FilterCondition]
+Param: TypeAlias = ProjectRowId
+
+Key: TypeAlias = Tuple[Kind, ProjectRowId, Optional[TimeRange], FilterCondition]
+Result: TypeAlias = SpanCount
+ResultPosition: TypeAlias = int
+DEFAULT_VALUE = 0
+
+
+class RecordCountDataLoader(DataLoader[Key, Result]):
+    def __init__(self, db: Callable[[], AsyncContextManager[AsyncSession]]) -> None:
+        super().__init__(load_fn=self._load_fn, cache_key_fn=self._cache_key_fn)
+        self._db = db
+
+    @staticmethod
+    def _cache_key_fn(key: Key) -> Tuple[Segment, Param]:
+        kind, project_rowid, time_range, filter_condition = key
+        interval = (
+            (time_range.start, time_range.end)
+            if isinstance(time_range, TimeRange)
+            else (None, None)
+        )
+        return (kind, interval, filter_condition), project_rowid
+
+    async def _load_fn(self, keys: List[Key]) -> List[Result]:
+        results: List[Result] = [DEFAULT_VALUE] * len(keys)
+        arguments: DefaultDict[
+            Segment,
+            DefaultDict[Param, List[ResultPosition]],
+        ] = defaultdict(lambda: defaultdict(list))
+        for position, key in enumerate(keys):
+            segment, param = self._cache_key_fn(key)
+            arguments[segment][param].append(position)
+        async with self._db() as session:
+            for segment, params in arguments.items():
+                stmt = _get_stmt(segment, *params.keys())
+                data = await session.stream(stmt)
+                async for project_rowid, span_count in data:
+                    for position in params[project_rowid]:
+                        results[position] = span_count
+        return results
+
+
+def _get_stmt(
+    segment: Segment,
+    *project_rowids: Param,
+) -> Select[Any]:
+    kind, (start_time, end_time), filter_condition = segment
+    pid = models.Trace.project_rowid
```

Comment:
Suggestion: `pid` has a different meaning, `project_id` probably better.

Reply:
On the other hand, `project_id` would at first glance suggest a scalar value like `123`, so it's also not ideal. And even though `pid` has other meanings, because the local context is highly specialized, it's unlikely to be misinterpreted. Furthermore, `pid` is succint and memorable.

When something has sufficient regularity and locality, i.e. sufficient repetition on a localized context, I think we should abbreviate it for one very good reason: by "anonymizing" the regular things, we would be better at accentuating the other, non-regular and unfamiliar things, which is what we actually want.